pyro/burgers/simulation.py

```diff
@@ -0,0 +1,163 @@
+import importlib
+
+import matplotlib
+import matplotlib.pyplot as plt
+import numpy as np
+
+import pyro.burgers.advective_fluxes as flx
+import pyro.mesh.patch as patch
+import pyro.particles.particles as particles
+import pyro.util.plot_tools as plot_tools
+from pyro.simulation_null import NullSimulation, bc_setup, grid_setup
+
+
+class Simulation(NullSimulation):
+
+    def initialize(self):
+        """
+        Initialize the grid and variables for advection and set the initial
+        conditions for the chosen problem.
+        """
+
+        # create grid, self.rp contains mesh.nx and mesh.ny
+        my_grid = grid_setup(self.rp, ng=4)
+
+        # create the variables
+        my_data = patch.CellCenterData2d(my_grid)
+
+        # outputs: bc, bc_xodd and bc_yodd for reflection boundary cond
+        bc = bc_setup(self.rp)[0]
+
+        # register variables in the data
+        # burgers equation advects velocity
+
+        my_data.register_var("x-velocity", bc)
+        my_data.register_var("y-velocity", bc)
+        my_data.create()
+
+        # holds various data, like time and registered variable.
+        self.cc_data = my_data
+
+        if self.rp.get_param("particles.do_particles") == 1:
+            n_particles = self.rp.get_param("particles.n_particles")
+            particle_generator = self.rp.get_param("particles.particle_generator")
+            self.particles = particles.Particles(self.cc_data, bc, n_particles, particle_generator)
+
+        # now set the initial conditions for the problem
+        problem = importlib.import_module(f"pyro.burgers.problems.{self.problem_name}")
+        problem.init_data(self.cc_data, self.rp)
+
+    def method_compute_timestep(self):
+        """
+        Compute the advective timestep (CFL) constraint.  We use the
+        driver.cfl parameter to control what fraction of the CFL
+        step we actually take.
+        """
+
+        cfl = self.rp.get_param("driver.cfl")
+
+        # since velocity is no longer a constant
+        # velocity varies in each zone
+
+        u = self.cc_data.get_var("x-velocity")
+        v = self.cc_data.get_var("y-velocity")
+
+        # dt = min(min(dx/|u_i|), min(dy/|v_j|))
+
+        xtmp = max(np.max(np.abs(u)), self.SMALL) / self.cc_data.grid.dx
+        ytmp = max(np.max(np.abs(v)), self.SMALL) / self.cc_data.grid.dy
+
+        self.dt = cfl / (xtmp + ytmp)
```

Comment:
I am not sure that this is the correct timestep limiter.  I think that we should essentially do the same thing that is done in the CTU advection or compressible solver